### PR TITLE
bugFix/snapshot-only-for-meta-epoch-start-data

### PR DESCRIPTION
**Pull request by sasurobert**
Snapshot only for meta epoch start data for metachain and shard as well.
If process local storage and shuffled out and synchronization failed, no need to retry the full sync again for start in epoch.
Fix epoch start meta header resolver.
Fix snapshot prunning buffer length.

### CODE REVIEW

File: epochStart/bootstrap/syncEpochStartMeta.go [hunk 1]
```diff
@@ -155,6 +155,11 @@ func (e *epochStartMetaSyncer) initTopicForEpochStartMetaBlockInterceptor() erro
 		return err
 	}
 
+	err = e.messenger.UnregisterMessageProcessor(factory.MetachainBlocksTopic)
```
Comment (SebastianMarian) [inline epochStart/bootstrap/syncEpochStartMeta.go:158]: This is sufficient for the scope of this PR?

Comment (sasurobert) [reply epochStart/bootstrap/syncEpochStartMeta.go]: forgot to push.

File: process/block/metablock.go [hunk 2]
```diff
@@ -1185,6 +1185,12 @@ func (mp *metaProcessor) updateState(lastMetaBlock data.HeaderHandler) {
 		return
 	}
 
+	if lastMetaBlock.IsStartOfEpochBlock() {
+		log.Debug("trie snapshot", "rootHash", lastMetaBlock.GetRootHash())
+		mp.accountsDB[state.UserAccountsState].SnapshotState(lastMetaBlock.GetRootHash())
+		mp.accountsDB[state.PeerAccountsState].SnapshotState(lastMetaBlock.GetRootHash())
```
Comment (BeniaminDrasovean) [inline process/block/metablock.go:1191]: lastMetaBlock.GetValidatorStatsRootHash() instead of lastMetaBlock.GetRootHash()

Comment (sasurobert) [reply process/block/metablock.go:1191]: thanks.

File: process/block/metablock.go [hunk 3]
```diff
@@ -1185,6 +1185,12 @@ func (mp *metaProcessor) updateState(lastMetaBlock data.HeaderHandler) {
 		return
 	}
 
+	if lastMetaBlock.IsStartOfEpochBlock() {
```
Comment (SebastianMarian) [inline process/block/metablock.go:1192]: Also here this snapshot could be executed on the same root hash more than once if a roll back occurs.

File: process/block/baseProcess.go [hunk 4]
```diff
@@ -979,11 +979,6 @@ func (bp *baseProcessor) updateStateStorage(
 		return
 	}
 
-	if finalHeader.IsStartOfEpochBlock() {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply process/block/baseProcess.go:993]: If there is a snapshot twice at the same root hash, the second snapshot will not be completed, because there already is a snapshot at that root hash, so there should be no problem.

File: process/block/metablock.go (hunk 3 above)
Comment (BeniaminDrasovean) [reply process/block/metablock.go:1192]: The second time the snapshot for the same root hash occurs, no snapshot will be done.

File: process/block/baseProcess.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/block/baseProcess.go:993]: Also you can create this map as following: map[rootHash][header nonce], and header nonce would be the nonce of the currentHeader param from func (sp *shardProcessor) updateState(headers []data.HeaderHandler, currentHeader *block.Header). In this way you can clean this map for nonces behind final  on CommitBlock method

[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/block/baseProcess.go:993]: The problem in snapShotEpochStartFromMeta appears also when a epoch start meta block appears in more shard headers which is possible as it could be fully processed in more iterations. So the above proposed protection will solve also this situation.

[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/block/baseProcess.go:993]: In case of shardBlockProcessor the nonce of this local map proposed above should be nonce of the start of epoch meta block and should be clean on CommitBlock if it is behind lastCrossNotarizedHeader nonce

[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/block/baseProcess.go:993]: In cse of metaBlockProcessor this nonce should be the nonce of the lastMetaBlock parameter from the updateState method and also should be clean on CommitBlock if it is behind highet final block nonce

[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/block/baseProcess.go:993]: Ok, cool, I have not seen the comment of @BeniaminDrasovean so forget about all these messages in this case :))

File: process/smartContract/process.go [hunk 5]
```diff
@@ -339,7 +339,9 @@ func (sc *scProcessor) resolveBuiltInFunctions(
 	}
 
 	scrResults := make([]data.TransactionHandler, 0, len(vmOutput.OutputAccounts)+1)
-	for _, outAcc := range vmOutput.OutputAccounts {
+
+	outPutAccounts := sortVMOutputInsideData(vmOutput)
```
Comment (SebastianMarian) [inline process/smartContract/process.go:343]: outputAccounts

Comment (sasurobert) [reply process/smartContract/process.go:343]: done.

File: epochStart/metachain/epochStartData.go [hunk 6]
```diff
@@ -332,7 +332,7 @@ func (e *epochStartData) computePendingMiniBlockList(
 
 		metaHdr, err := e.getMetaBlockByHash(shardData.FirstPendingMetaBlock)
 		if err != nil {
-			go e.requestHandler.RequestMetaHeader(shardData.FirstPendingMetaBlock)
+			e.requestHandler.RequestMetaHeader(shardData.FirstPendingMetaBlock)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/metachain/epochStartData.go:335]: there was no problem

File: epochStart/metachain/epochStartData.go [hunk 7]
```diff
@@ -209,7 +209,7 @@ func (e *epochStartData) lastFinalizedFirstPendingListHeadersForShard(shardHdr *
 	for currentHdr := shardHdr; currentHdr.GetNonce() > 0 && currentHdr.GetEpoch() == shardHdr.GetEpoch(); {
 		prevShardHdr, err := process.GetShardHeader(currentHdr.GetPrevHash(), e.dataPool.Headers(), e.marshalizer, e.store)
 		if err != nil {
-			go e.requestHandler.RequestShardHeader(currentHdr.ShardID, currentHdr.GetPrevHash())
+			e.requestHandler.RequestShardHeader(currentHdr.ShardID, currentHdr.GetPrevHash())
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/metachain/epochStartData.go:212]: there was no problem

File: process/smartContract/process.go [hunk 8]
```diff
@@ -294,7 +295,8 @@ func (sc *scProcessor) deleteSCRsWithValueZeroGoingToMeta(scrs []data.Transactio
 		}
 		cleanSCRs = append(cleanSCRs, scr)
 	}
-	return cleanSCRs
+
+	return transaction.TrimSliceHandler(cleanSCRs)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/process.go:299]: it is not necessary.

File: data/trie/trieStorageManager.go [hunk 9]
```diff
@@ -352,9 +352,6 @@ func (tsm *trieStorageManager) writeOnChan(entry snapshotsQueueEntry) {
 	select {
 	case tsm.snapshotReq <- entry:
 		return
-	default:
```
Comment (SebastianMarian) [inline data/trie/trieStorageManager.go:355]: Maybe we should let this print and change the value of SnapshotsBufferLen from config.toml to something more reasonable as 1000 and if something goes wrong and these messages would not be consumed we could be triggered by this print.

File: cmd/node/config/economics.toml [hunk 10]
```diff
@@ -1,8 +1,8 @@
 #Economics config of the node
 [GlobalSettings]
-    TotalSupply = "20000000000000000000000000000" #20BILERD
+    TotalSupply = "200000000000000000000000000000" #20BILERD
```
Comment (SebastianMarian) [inline cmd/node/config/economics.toml:3]: Now are 200BIL + 18 zecimals

File: epochStart/metachain/epochStartData.go [hunk 11]
```diff
@@ -210,11 +208,8 @@ func (e *epochStartData) lastFinalizedFirstPendingListHeadersForShard(shardHdr *
 		prevShardHdr, err := process.GetShardHeader(currentHdr.GetPrevHash(), e.dataPool.Headers(), e.marshalizer, e.store)
 		if err != nil {
 			go e.requestHandler.RequestShardHeader(currentHdr.ShardID, currentHdr.GetPrevHash())
-			if e.epochStartTrigger.Epoch()-currentHdr.GetEpoch() > maxEpochDifference {
-				log.Warn("shard remained in an epoch that is too old", "shardID", currentHdr.ShardID, "shard Epoch", currentHdr.Epoch, "meta Epoch", e.epochStartTrigger.Epoch())
-				break
-			}
-			return nil, nil, nil, err
+			log.Warn("shard remained in an epoch that is too old", "shardID", currentHdr.ShardID, "shard Epoch", currentHdr.Epoch, "meta Epoch", e.epochStartTrigger.Epoch())
```
Comment (iulianpascalau) [inline epochStart/metachain/epochStartData.go:211]: Is it ok to leave this warn? It seems unrelated now. Why not logging also the real error err

File: cmd/node/config/economics.toml [hunk 12]
```diff
@@ -13,20 +13,20 @@
...
     GasPerDataByte = "1500"
     DataLimitForBaseCalc = "10000"
 
 [ValidatorSettings]
-    GenesisNodePrice = "500000000000000000000000" #500000ERD
+    GenesisNodePrice = "2500000000000000000000000" #2.5MILERD
     UnBondPeriod = "400"
     MinStepValue = "10000000000000000000000"
-    TotalSupply = "20000000000000000000000000000" #20BIL ERD
+    TotalSupply = "200000000000000000000000000000" #20BIL ERD
```
Comment (SebastianMarian) [inline cmd/node/config/economics.toml:25]: Now are 200BIL + 18 zecimals

File: cmd/node/config/economics.toml (hunk 10 above)
Comment (iulianpascalau) [inline cmd/node/config/economics.toml:3]: You have an extra 0. What is wrote there is actually 200 bil

File: cmd/node/config/economics.toml (hunk 12 above)
Comment (iulianpascalau) [inline cmd/node/config/economics.toml:25]: also here, there is one extra 0

File: cmd/node/config/economics.toml [hunk 13]
```diff
@@ -13,20 +13,20 @@
...
     AuctionEnableNonce = "100000000000"
     StakeEnableNonce = "0"
     NumRoundsWithoutBleed = "100"
     MaximumPercentageToBleed = "0.5"
     BleedPercentagePerRound = "0.00001"
-    UnJailValue = "500000000000000000000"
+    UnJailValue = "250000000000000000000" #0.1% of genesis node price
```
Comment (iulianpascalau) [inline cmd/node/config/economics.toml:31]: In order to be 0.1%, you need an extra 0 as the value there is 250 ERD

File: epochStart/metachain/epochStartData.go (hunk 11 above)
Comment (sasurobert) [reply epochStart/metachain/epochStartData.go:211]: the real error is that it does not find the header.

File: cmd/node/config/economics.toml (hunk 10 above)
Comment (sasurobert) [reply cmd/node/config/economics.toml:3]: yep.

File: cmd/node/config/economics.toml (hunk 12 above)
Comment (sasurobert) [reply cmd/node/config/economics.toml:25]: done

File: cmd/node/config/economics.toml (hunk 13 above)
Comment (sasurobert) [reply cmd/node/config/economics.toml:31]: done.

File: cmd/node/config/economics.toml [hunk 14]
```diff
@@ -13,20 +13,20 @@
 
 [FeeSettings]
     MaxGasLimitPerBlock = "1500000000"
-    MinGasPrice = "100000000000000" #will yield min tx fee of 10ERD
+    MinGasPrice = "200000000000" #will yield min tx fee of 0.01ERD
```
Comment (SebastianMarian) [inline cmd/node/config/economics.toml:16]: This means 0.002ERD applying DataLimitForBaseCalc 10^4

File: cmd/node/config/economics.toml [hunk 15]
```diff
@@ -13,20 +13,20 @@
...
     MinStepValue = "10000000000000000000000"
-    TotalSupply = "20000000000000000000000000000" #20BIL ERD
+    TotalSupply = "200000000000000000000000000000" #20BIL ERD
     AuctionEnableNonce = "100000000000"
     StakeEnableNonce = "0"
     NumRoundsWithoutBleed = "100"
     MaximumPercentageToBleed = "0.5"
     BleedPercentagePerRound = "0.00001"
-    UnJailValue = "50000000000000000000" #0.1% of genesis node price
+    UnJailValue = "250000000000000000000" #0.1% of genesis node price
```
Comment (SebastianMarian) [inline cmd/node/config/economics.toml:31]: This is now 0.01% from genesis node price (4 of 0 less)

File: dataRetriever/mock/epochHandlerStub.go [hunk 16]
```diff
@@ -2,13 +2,13 @@ package mock
 
 // EpochHandlerStub -
 type EpochHandlerStub struct {
-	EpochCalled func() uint32
+	MetaEpochCalled func() uint32
 }
 
 // Epoch -
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/mock/epochHandlerStub.go:8]: done.

File: dataRetriever/resolvers/epochproviders/nilEpochHandler.go [hunk 17]
```diff
@@ -9,7 +9,7 @@ func NewNilEpochHandler() *nilEpochHandler {
 }
 
 // Epoch returns the current epoch
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/resolvers/epochproviders/nilEpochHandler.go:11]: done

File: epochStart/metachain/epochStartData.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/metachain/epochStartData.go:211]: in case of meta it is the same. searched in the whole project.

File: data/trie/trieStorageManager.go (hunk 9 above)
Comment (sasurobert) [reply data/trie/trieStorageManager.go:355]: there is no need for this default. 1000 is not enough. you can have much more data tries in the same shard.

File: cmd/node/config/economics.toml (hunk 10 above)
Comment (sasurobert) [reply cmd/node/config/economics.toml]: done

File: cmd/node/config/economics.toml (hunk 12 above)
Comment (sasurobert) [reply cmd/node/config/economics.toml]: done

File: cmd/node/config/economics.toml (hunk 14 above)
Comment (sasurobert) [reply cmd/node/config/economics.toml:16]: MinGasPrice * MinGasLimit - gaslimit of a transaction is 50.000

File: cmd/node/config/economics.toml (hunk 15 above)
Comment (sasurobert) [reply cmd/node/config/economics.toml]: done.

File: epochStart/metachain/epochStartData.go (hunk 11 above)
Comment (iulianpascalau) [reply epochStart/metachain/epochStartData.go]: so the message "shard remained in an epoch that is too old" is relevant?

File: data/trie/trieStorageManager.go (hunk 9 above)
Comment (SebastianMarian) [reply data/trie/trieStorageManager.go:355]: ok

File: epochStart/metachain/epochStartData.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply epochStart/metachain/epochStartData.go]: ok

File: integrationTests/multiShard/endOfEpoch/startInEpoch/startInEpoch_test.go [hunk 18]
```diff
@@ -213,6 +214,7 @@ func testNodeStartsInEpoch(t *testing.T, shardID uint32, expectedHighestRound ui
 	epochStartBootstrap, err := bootstrap.NewEpochStartBootstrap(argsBootstrapHandler)
 	assert.Nil(t, err)
 
+	_ = logger.SetLogLevel("*:DEBUG")
```
Comment (SebastianMarian) [inline integrationTests/multiShard/endOfEpoch/startInEpoch/startInEpoch_test.go:217]: Do you need this log to remain here?

Comment (sasurobert) [reply integrationTests/multiShard/endOfEpoch/startInEpoch/startInEpoch_test.go:217]: no but it was a good log to find the problem in the test.
